fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/UnequalPredicateInfer.java

```diff
@@ -0,0 +1,532 @@
+                        continue;
+                    } else if (!(inputExprs.get(i) instanceof Literal) && !(inputExprs.get(j) instanceof Literal)) {
+                        if (isTableFilter(i, j)) {
+                            tableFilters.add(Pair.of(i, j));
+                        } else {
+                            nonTableFilters.add(Pair.of(i, j));
+                        }
+                    } else if (inputExprs.get(i) instanceof Literal
+                            || inputExprs.get(j) instanceof Literal) {
+                        set(chosen, i, j, Relation.EQ);
+                        if (inputExprs.get(i) instanceof Literal) {
+                            equalToLiteral[j] = i;
+                            equalWithConstant.add(j);
+                        } else {
+                            equalToLiteral[i] = j;
+                            equalWithConstant.add(i);
+                        }
+                    }
+                }
+            }
+            // a=b a=c a=1 only infer a=1 b=1 c=1, not retain a=b a=c
+            for (Pair<Integer, Integer> tableFilter : tableFilters) {
+                int left = tableFilter.first;
+                int right = tableFilter.second;
+                if (equalToLiteral[left] == -1 || equalToLiteral[right] == -1) {
+                    set(chosen, left, right, Relation.EQ);
+                    equalToLiteral[left] = left;
+                    equalToLiteral[right] = left;
+                }
+            }
+            for (Pair<Integer, Integer> nonTableFilter : nonTableFilters) {
+                int left = nonTableFilter.first;
+                int right = nonTableFilter.second;
+                if (!equalWithConstant.contains(left) && !equalWithConstant.contains(right)) {
+                    set(chosen, left, right, Relation.EQ);
+                }
+            }
+            return chosen;
+        }
+
+        private Expression normalize(ComparisonPredicate cmp) {
+            return TypeCoercionUtils.processComparisonPredicate(normalizePredicate(cmp)).withInferred(true);
+        }
+
+        /** for test */
+        public Relation[][] getGraph() {
+            return graph;
+        }
+    }
+
+    /**inferUnequalPredicates*/
+    public static Set<? extends Expression> inferUnequalPredicates(Set<ComparisonPredicate> inputs) {
+        if (inputs.size() < 2) {
+            return inputs;
+        }
+        InferenceGraph inferGraph = new InferenceGraph(inputs);
+        if (inferGraph.inputExprs.isEmpty()) {
+            return inputs;
+        }
+        inferGraph.deduce(inferGraph.graph);
+        Set<Integer> equalWithConstant = new HashSet<>();
+        InferenceGraph.Relation[][] chosen = inferGraph.chooseEqualPredicates(equalWithConstant);
+        inferGraph.chooseUnequalPredicates(chosen, equalWithConstant);
+        Set<Expression> newPredicates = inferGraph.chooseInputPredicates(chosen);
```

Comment:
pls record a predicates' order-preserving jira later and measure the related performance impact.